Solutions/Azure Active Directory/Analytic Rules/AdminPromoAfterRoleMgmtAppPermissionGrant.yaml

```diff
@@ -24,54 +24,73 @@ relevantTechniques:
-  | extend Target = tostring(parse_json(tostring(TargetResources.modifiedProperties[4].newValue)))
-  | extend TargetId = tostring(parse_json(tostring(TargetResources.modifiedProperties[3].newValue)))
-  | project TimeGenerated, OperationName, Initiator, Target, TargetId, Result
-  | join kind=innerunique (
-    AuditLogs
-    | where LoggedByService =~ "Core Directory"
-    | where Category =~ "RoleManagement"
-    | where AADOperationType in ("Assign", "AssignEligibleRole")
-    | where ActivityDisplayName has_any ("Add eligible member to role", "Add member to role")
-    | mv-expand TargetResources
-    | mv-expand TargetResources.modifiedProperties
-    | extend displayName_ = tostring(TargetResources_modifiedProperties.displayName)
-    | where displayName_ =~ "Role.DisplayName"
-    | extend RoleName = tostring(parse_json(tostring(TargetResources_modifiedProperties.newValue)))
-    | where RoleName contains "Admin"
-    | extend Initiator = tostring(parse_json(tostring(InitiatedBy.app)).displayName)
-    | extend InitiatorId = tostring(parse_json(tostring(InitiatedBy.app)).servicePrincipalId)
-    | extend TargetUser = tostring(TargetResources.userPrincipalName)
-    | extend Target = iif(isnotempty(TargetUser), TargetUser, tostring(TargetResources.displayName))
-    | extend TargetType = tostring(TargetResources.type)
-    | extend TargetId = tostring(TargetResources.id)
-    | project TimeGenerated, OperationName,  RoleName, Initiator, InitiatorId, Target, TargetId, TargetType, Result
-  ) on $left.TargetId == $right.InitiatorId
-  | extend TimeRoleMgGrant = TimeGenerated, TimeAdminPromo = TimeGenerated1, ServicePrincipal = Initiator1, ServicePrincipalId = InitiatorId,
-    TargetObject = Target1, TargetObjectId = TargetId1, TargetObjectType = TargetType
-  | where TimeRoleMgGrant < TimeAdminPromo
-  | project TimeRoleMgGrant, TimeAdminPromo, RoleName, ServicePrincipal, ServicePrincipalId, TargetObject, TargetObjectId, TargetObjectType
+  | where TimeGenerated > ago(query_period)
+  | where Category =~ "ApplicationManagement" and LoggedByService =~ "Core Directory"
+  | where OperationName =~ "Add app role assignment to service principal"
+  | mv-expand TargetResource = TargetResources
+  | mv-expand modifiedProperty = TargetResource["modifiedProperties"]
+  | where tostring(modifiedProperty["displayName"]) == "AppRole.Value"
+  | extend PermissionGrant = tostring(modifiedProperty["newValue"])
+  | where PermissionGrant has "RoleManagement.ReadWrite.Directory"
+  | mv-apply modifiedProperty = TargetResource["modifiedProperties"] on (
+      summarize modifiedProperties = make_bag(
+          pack(tostring(modifiedProperty["displayName"]),
+              pack("oldValue", trim(@'[\"\s]+', tostring(modifiedProperty["oldValue"])),
+                  "newValue", trim(@'[\"\s]+', tostring(modifiedProperty["newValue"])))))
```

Comment:
1. pack function is deprecated, use bag_pack instead.
2. As a best practice, make_bag function should have maxSize parameter. Assign a sufficient large maxSize parameter (100 seems big enough).